File: node/src/webhook.ts
import { createHmac, timingSafeEqual } from 'crypto';

export const RESPATCH_SIGNATURE_HEADER = 'X-Respatch-Signature';

/** @deprecated Use RESPATCH_SIGNATURE_HEADER */
export const TRANSMIT_SIGNATURE_HEADER = RESPATCH_SIGNATURE_HEADER;

/**
 * Verify an inbound Respatch webhook using `X-Respatch-Signature` (or legacy `X-Transmit-Signature`).
 * Signature is HMAC-SHA256 of the raw request body, hex-encoded.
 */
export function verifyWebhookSignature(
  payload: string | Buffer,
  signatureHeader: string,
  secret: string,
): boolean {
  if (!signatureHeader || !secret) return false;

  const body = typeof payload === 'string' ? payload : payload.toString('utf8');
  const expected = createHmac('sha256', secret).update(body, 'utf8').digest('hex');
  const provided = signatureHeader.replace(/^sha256=/i, '').trim();

  try {
    const a = Buffer.from(expected, 'hex');
    const b = Buffer.from(provided, 'hex');
    return a.length === b.length && timingSafeEqual(a, b);
  } catch {
    return false;
  }
}

/** Resolve signature header from incoming request headers (Respatch preferred). */
export function getWebhookSignatureHeader(
  headers: Record<string, string | string[] | undefined>,
): string {
  const respatch = headers[RESPATCH_SIGNATURE_HEADER.toLowerCase()];
  if (typeof respatch === 'string') return respatch;
  const legacy = headers['x-transmit-signature'];
  if (typeof legacy === 'string') return legacy;
  return '';
}

export function signWebhookPayload(payload: string | Buffer, secret: string): string {
  const body = typeof payload === 'string' ? payload : payload.toString('utf8');
  return createHmac('sha256', secret).update(body, 'utf8').digest('hex');
}

File: rust/src/webhook.rs
use hmac::{Hmac, Mac};
use sha2::Sha256;

type HmacSha256 = Hmac<Sha256>;

/// Verify `X-Transmit-Signature` (HMAC-SHA256 hex of raw body).
pub fn verify_webhook_signature(payload: &[u8], signature_header: &str, secret: &str) -> bool {
    if signature_header.is_empty() || secret.is_empty() {
        return false;
    }

    let provided = signature_header
        .trim()
        .strip_prefix("sha256=")
        .unwrap_or(signature_header.trim());

    let mut mac = match HmacSha256::new_from_slice(secret.as_bytes()) {
        Ok(m) => m,
        Err(_) => return false,
    };
    mac.update(payload);

    let expected = hex::encode(mac.finalize().into_bytes());
    constant_time_eq(expected.as_bytes(), provided.as_bytes())
}

/// Sign a payload for tests or replay tools.
pub fn sign_webhook_payload(payload: &[u8], secret: &str) -> String {
    let mut mac = HmacSha256::new_from_slice(secret.as_bytes()).expect("hmac key");
    mac.update(payload);
    hex::encode(mac.finalize().into_bytes())
}

fn constant_time_eq(a: &[u8], b: &[u8]) -> bool {
    if a.len() != b.len() {
        return false;
    }
    a.iter()
        .zip(b.iter())
        .fold(0u8, |acc, (x, y)| acc | (x ^ y))
        == 0
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn round_trip_signature() {
        let secret = "whsec_test";
        let body = br#"{"event":"delivery.status_changed"}"#;
        let sig = sign_webhook_payload(body, secret);
        assert!(verify_webhook_signature(body, &sig, secret));
        assert!(verify_webhook_signature(body, &format!("sha256={sig}"), secret));
    }
}
